TestNewBinaryIOCleanup failing on gotip

### Description

Noticed that there is a test failure on gotip (future Go 1.23):

```
=== RUN   TestNewBinaryIOCleanup
    io_test.go:64: some descriptors weren't closed (7 != 8)
--- FAIL: TestNewBinaryIOCleanup (0.00s)
FAIL
FAIL    github.com/containerd/containerd/v2/cmd/containerd-shim-runc-v2/process 0.003s
FAIL
```

This happens because 1.23 will make use of pidfd on Linux, see https://github.com/golang/go/issues/62654.

The test thinks that this file descriptor is leaking. The solution is pretty simple, it needs to ignore a file descriptor that is pidfd.

This is a hacky way to ignore file descriptors internal to the runtime, there is a proposal open at https://github.com/golang/go/issues/67639 that should provide a better way to detect leaking fds.

### Steps to reproduce the issue

```
$ go install golang.org/dl/gotip@latest
$ gotip download
$ gotip test -v ./cmd/containerd-shim-runc-v2/process -run=TestNewBinaryIOCleanup
```


### Describe the results you received and expected

Test fails. It should pass and continue to work on 1.21 and 1.22 as well.

### What version of containerd are you using?

tip

### Any other relevant information

_No response_

### Show configuration if it is related to CRI plugin.

_No response_

--- a/cmd/containerd-shim-runc-v2/process/io_test.go
+++ b/cmd/containerd-shim-runc-v2/process/io_test.go
@@ -20,8 +20,10 @@ package process
 
 import (
 	"context"
+	"fmt"
 	"net/url"
 	"os"
+	"strings"
 	"testing"
 
 	"github.com/containerd/containerd/v2/pkg/namespaces"
@@ -68,5 +70,28 @@ func TestNewBinaryIOCleanup(t *testing.T) {
 func descriptorCount(t *testing.T) int {
 	t.Helper()
 	files, _ := os.ReadDir("/proc/self/fd")
+
+	// Go 1.23 introduced a new internal file descriptor type "pidfd"
+	// that we don't want to count towards the total file descriptors in
+	// use by the process. This retains the behavior of previous Go
+	// versions.
+	// See https://go.dev/issues/62654.
+	//
+	// Once the proposal to check for internal file descriptors is
+	// accepted, we can use that instead to detect internal fds in use
+	// by the Go runtime.
+	// See https://go.dev/issues/67639.
+	for i, file := range files {
+		sym, err := os.Readlink(fmt.Sprintf("/proc/self/fd/%s", file.Name()))
+		if err != nil {
+			// ignore fds that cannot be followed.
+			continue
+		}
+
+		if strings.HasPrefix(sym, "pidfd:") {
+			files = append(files[:i], files[i+1:]...)
+		}
+	}
+
 	return len(files)
 }

